xwiki-platform-core/xwiki-platform-web/xwiki-platform-web-war/src/main/webapp/resources/js/xwiki/xwiki.js

```diff
@@ -1992,3 +1992,54 @@ document.observe("xwiki:dom:loaded", function() {
   interceptXMLHttpRequest();
   interceptFetch();
 })();
+
+/**
+ * Initializes the submit-once protection on the current page.
+ * Search for all form elements with a .xwiki-submit-once class one the document (or on an updated sub-part of the dom).
+ * For each form, once the form is submitted, mark it with the .xwiki-form-submitting class and prevent new submissions.
```

Comment:
Can't we just disable the form / buttons to prevent re-submission? (back in the day this used to be the classical solution for this problem)

And then, the visual feedback that a user would get about this would be the visual feedback of "this is not clickable", which would be coherent with all the other forms that would not be clickable (regardless of the reason for which they're not clickable...).

Reply:
> This solution looks a little complex to me, can't we just disable the form / buttons to prevent re-submission? (back in the day this used to be the classical solution for this problem)

You can't disable the fields, their values wouldn't be sent server side.
I proposed to mark them are `readonly` which would lead more or less to the same result, but does not prevent buttons to be clicked.
Also, @mflorea rightfully suggested that it was best if we could avoid updating the whole form, and only update a single class at the root of the form.

For the UX pov I agree with you but I have no idea how to achieve this.

Reply:
> You can't disable the fields, their values wouldn't be sent server side.

the fields don't need to be disabled. The submit button needs to be disabled.

Since it's just a submit button, the server side management of the form should not depend on the presence of the button itself on the request.

I had the feeling that disabling submit buttons when a POST form is submitted was a sort of a best practice since forever (there are a bunch of examples in this direction on the internet, but indeed there are also many in the direction you implemented, with classes & all).

Reply:
> Since it's just a submit button, the server side management of the form should not depend on the presence of the button itself on the request.

This is not exactly correct. With the form below, the value of `a` will depend on the button that trigger the form submit:
```html
<form method="get" action="index.html">
  <input type="submit" name="a" value="1">
  <input type="submit" name="a" value="2">
  <button type="submit" name="a" value="3">XXX</button>
</form>
```

You can try with and without the following js code. The `a` param is missing when the js is here.
```
$(() => {
      let form = $("form");
      form.submit(() => {
        form.find(':input').prop('disabled', true);
      });
    });
```

Then, we can decide to provide a first solution that would be limited to form where the button values can be safely ignored. But that would be less general.

Reply:
> This is not exactly correct. With the form below, the value of a will depend on the button that trigger the form submit:

yes, I didn't mean for the general case where there are multiple submit buttons with different values, I meant for this specific form and generally for all forms with a single submit button - even if I know that there _may_ be forms with a single submit button where the submit value is used on the server side, I consider them rare.

> Then, we can decide to provide a first solution that would be limited to form where the button values can be safely ignored. But that would be less general.

I think it would make sense to try to go for a solution that is less generic (at least for the forms that have a single submit button whose value is not needed on the server), as going for the very generic case makes it _too_ generic and may make it complex and long to implement way beyond the need of the login form (it's a typical case where "best" kills "better").
For example, if we want to take into account the _very_ generic usecase of a form with multiple submit buttons, we'd need to also define what blocking a double submit would mean: would it mean that the same button cannot be clicked twice or would it mean that no other button can be clicked a second time?

Reply:
I have updated the PR with a simpler version where the submit buttons of the submitted form are disable during  the first submit.

Reply:
It's not clear to me what was wrong with the initial solution and how is the new solution better.

> The complexity, for me, comes from the fact that we take the responsibility to prevent clicks in our own code, while click preventing is something that already exists as a feature of html, it's called disabled forms / buttons.

@lucaa maybe I missed something, but I don't remember seeing any **click** listener. There was just a **submit** listener which is still present. The only difference I see is that now we also disable **all** the submit buttons. What's the benefit of this? I can only see the downside: the submit button value won't be sent to the server.

Reply:
> There was just a submit listener which is still present. 

Yeah, that's what I was talking about. There was some logic associated to preventing it, which is still indeed there in the new PR and it's redundant from my pov. I commented on the commit.

Reply:
> It's not clear to me what was wrong with the initial solution and how is the new solution better.

to clarify this too: what I found "too much" in the initial solution is that we were coding the logic of keeping track of the state of the form (submitted, not submitted yet) and also accepting additional form submissions in the logic of our code, only to stop them (also in our code) if the state of the form says that it was already submitted. 
What I was expecting by using the standard disabling mechanism was something along the lines of: when a submit happens, "turn off" the possibility to submit form and just forget about it, nothing else can happen further, we don't need to worry about what will happen further. 

More like instead of allowing an event to happen more than once and then determining, every time it happens, whenther we should let it happen or not based on a state we stored, we just "turn off" its source when it happens first and we don't need to think about to do when it happens again.